snuba_sdk/query_visitors.py

```diff
@@ -0,0 +1,321 @@
+from __future__ import annotations
+
+import json
+from abc import ABC, abstractmethod
+from typing import (
+    Generic,
+    List,
+    MutableMapping,
+    Optional,
+    TypeVar,
+    TYPE_CHECKING,
+    Union,
+)
+
+from snuba_sdk.entity import Entity
+from snuba_sdk.conditions import Condition
+from snuba_sdk.expressions import (
+    Column,
+    Function,
+    Granularity,
+    Limit,
+    LimitBy,
+    Offset,
+    OrderBy,
+    Totals,
+)
+from snuba_sdk.visitors import Translation
+
+if TYPE_CHECKING:
+    from snuba_sdk.query import Query
+
+
+class InvalidQuery(Exception):
+    pass
+
+
+QVisited = TypeVar("QVisited")
+
+
+class QueryVisitor(ABC, Generic[QVisited]):
+    def visit(self, query: Query) -> QVisited:
```

Comment:
I spent a bunch of time trying to figure out if I could get mypy to show an error if the fields of the query were modified. I'm still open to suggestions about how to do that.

Reply:
>  if I could get mypy to show an error if the fields of the query were modified

This is not possible. mypy does not act at runtime, thus there is no way for it to assert that a call to a method returns a different object with respect to the one you passed. It can only check types statically. 
The only way I know of to avoid that kind of issue (passing an object to a method and ensure it is not changed) is by making the object immutable and ensure mypy is used so any attempt of mutation is caught by the type checker. 
A less compelling alternatives is passing copies of the objects you want to ensure are not changed and discard the copy afterwards.

Reply:
I wasn't talking about at runtime, I meant a mypy compile error. If you are editing the code and you add a field to the query, you have to add it to a number of other functions and I wanted mypy to warn you of that.

Reply:
Ah so you were not talking about visit methods mutating the content of the argument they receive but the structure of the query changing by adding fields. Sorry my bad.
Then I am not sure about the problem. If you add a new field to the query, for as long as you add it to the abstract visitor and the visit method of the abstract calls it, all subclasses will need to implement that for mypy to be ok. (you may have to drop the idea of  dynamically building the method name you call since mypy cannot check that).
If instead the issue is in ensuring that `combine` methods take the new field into account, then there is no way neither with this architecture nor by adding everything in the query class. You can make it harder to miss by exploding the arguments to the combine method. Instead of passing an opaque dict (or event a namedtuple or dataclass), pass each individual argument. This forces you to touch that method in every visitor implementation and at least think about it.

Reply:
It's not a problem, I just thought it would be nice if when you add a new field to the query, mypy will warn you about all the other places you need to add the field.